File: src/parser/map.zig
const std = @import("std");
const ast = @import("../ast/map.zig");
const Parser = @import("parser.zig").Parser;

/// Parses a map declaration, e.g.:
/// map connection_counter {
///     type: .hash;
///     key: u32;
///     value: u64;
///     max: 1024;
/// }
pub fn parseMap(self: *Parser) Parser.ParseError!ast.MapDecl {
    const map_loc = self.current.loc;

    // Map name
    const map_name_tok = try self.expect(.identifier);
    _ = try self.expect(.l_brace);

    var map_type: ?ast.MapType = null;
    var key_type: ?ast.Type = null;
    var value_type: ?ast.Type = null;
    var max_entries: ?u32 = null;

    while (!self.check(.r_brace)) {
        // --- type ---
        if (self.match(.keyword_type)) {
            _ = try self.expect(.colon);

            _ = try self.expect(.dot);
            const t_tok = try self.expect(.identifier);
            // assign map_type using string comparison
            if (std.mem.eql(u8, t_tok.lexeme, "hash")) {
                map_type = .hash;
            } else if (std.mem.eql(u8, t_tok.lexeme, "array")) {
                map_type = .array;
            } else if (std.mem.eql(u8, t_tok.lexeme, "ringbuf")) {
                map_type = .ringbuf;
            } else if (std.mem.eql(u8, t_tok.lexeme, "lru_hash")) {
                map_type = .lru_hash;
            } else if (std.mem.eql(u8, t_tok.lexeme, "prog_array")) {
                map_type = .prog_array;
            } else {
                return self.parseError("expected map type");
            }

            _ = try self.expect(.semicolon); // expect ';'
            continue;
        }

        // --- key type ---
        if (self.match(.keyword_key)) {
            _ = try self.expect(.colon);
            key_type = try parseType(self);
            _ = try self.expect(.semicolon);
            continue;
        }

        // --- value type ---
        if (self.match(.keyword_value)) {
            _ = try self.expect(.colon);
            value_type = try parseType(self);
            _ = try self.expect(.semicolon);
            continue;
        }

        // --- max entries ---
        if (self.match(.keyword_max)) {
            _ = try self.expect(.colon);
            const n = try self.expect(.number);
            if (n.int_value < 0) return self.parseError("max must be >= 0");
            max_entries = @as(u32, @intCast(n.int_value));
            _ = try self.expect(.semicolon);
            continue;
        }

        // Unexpected token inside map
        return self.parseError("unexpected token in map");
    }

    // --- closing brace ---
    _ = try self.expect(.r_brace);

    // --- validate required fields ---
    if (map_type == null or key_type == null or value_type == null or max_entries == null) {
        return self.parseError("map missing required fields: type, key, value, max");
    }

    return .{
        .name = try self.allocator.dupe(u8, map_name_tok.lexeme),
        .map_type = map_type.?,
        .key_type = key_type.?,
        .value_type = value_type.?,
        .max_entries = max_entries.?,
        .loc = map_loc,
    };
}

/// Parses simple types for key/value fields
fn parseType(self: *Parser) Parser.ParseError!ast.Type {
    const t = self.current;
    _ = try self.advance();

    switch (t.kind) {
        .type_u32 => return .u32,
        .type_u64 => return .u64,
        .type_i32 => return .i32,
        .type_i64 => return .i64,
        else => return self.parseError("expected type"),
    }
}

File: src/parser/map.rs
use super::{Parser, ParseError};
use crate::{ast::{MapDecl, MapType, Type}, parser::TokenKind};

pub fn parse_map(parser: &mut Parser) -> Result<MapDecl, ParseError> {
    let map_loc = parser.current_loc();

    let map_name_tok = parser.expect(TokenKind::Identifier)?;
    expect_token(parser, TokenKind::LBrace)?;

    let mut map_type: Option<MapType> = None;
    let mut key_type: Option<Type> = None;
    let mut value_type: Option<Type> = None;
    let mut max_entries: Option<u32> = None;

    while !parser.check(TokenKind::RBrace) {
        if parser.r#match(TokenKind::KeywordType) {
            expect_token(parser, TokenKind::Colon)?;
            expect_token(parser, TokenKind::Dot)?;
            let t_tok = parser.expect(TokenKind::Identifier)?;

            map_type = Some(match t_tok.lexeme.as_str() {
                "hash" => MapType::Hash,
                "array" => MapType::Array,
                "ringbuf" => MapType::Ringbuf,
                "lru_hash" => MapType::LruHash,
                "prog_array" => MapType::ProgArray,
                _ => {
                    return Err(parser.error_with_help(
                        format!("Unknown map type: {}", t_tok.lexeme),
                        "Valid types: hash, array, ringbuf, lru_hash, prog_array"
                    ));
                }
            });

            expect_token(parser, TokenKind::Semicolon)?;
            continue;
        }
        if parser.r#match(TokenKind::KeywordKey) {
            expect_token(parser, TokenKind::Colon)?;
            key_type = Some(parse_type(parser)?);
            expect_token(parser, TokenKind::Semicolon)?;
            continue;
        }
        if parser.r#match(TokenKind::KeywordValue) {
            expect_token(parser, TokenKind::Colon)?;
            value_type = Some(parse_type(parser)?);
            expect_token(parser, TokenKind::Semicolon)?;
            continue;
        }        
        if parser.r#match(TokenKind::KeywordMax) {
            expect_token(parser, TokenKind::Colon)?;
            let n = parser.expect(TokenKind::Number)?;
            
            let max = n.int_value.ok_or_else(|| {
                parser.error("Expected integer value for max_entries")
            })?;
            
            if max < 0 {
                return Err(parser.error("max_entries must be >= 0"));
            }
            
            max_entries = Some(max as u32);
            expect_token(parser, TokenKind::Semicolon)?;
            continue;
        }
        
        return Err(parser.error_with_help(
            format!("Unexpected token inside map: {}", parser.current_kind()),
            "Expected one of: type, key, value, max"
        ));
    }
    
    expect_token(parser, TokenKind::RBrace)?;
    
    if map_type.is_none() {
        return Err(parser.error("Map missing required field: type"));
    }
    if key_type.is_none() {
        return Err(parser.error("Map missing required field: key"));
    }
    if value_type.is_none() {
        return Err(parser.error("Map missing required field: value"));
    }
    if max_entries.is_none() {
        return Err(parser.error("Map missing required field: max"));
    }

    Ok(MapDecl {
        name: map_name_tok.lexeme,
        map_type: map_type.unwrap(),
        key_type: key_type.unwrap(),
        value_type: value_type.unwrap(),
        max_entries: max_entries.unwrap(),
        loc: map_loc,
    })
}

pub fn parse_type(parser: &mut Parser) -> Result<Type, ParseError> {
    let t = parser.current().clone();
    parser.advance()?;

    match t.kind {
        TokenKind::TypeU32 => Ok(Type::U32),
        TokenKind::TypeU64 => Ok(Type::U64),
        TokenKind::TypeI32 => Ok(Type::I32),
        TokenKind::TypeI64 => Ok(Type::I64),
        _ => Err(parser.error("Expected type (u32, u64, i32, i64)")),
    }
}

pub fn expect_token(parser: &mut Parser, kind: TokenKind) -> Result<(), ParseError> {
    parser.expect(kind)?;
    Ok(())
}